src/validate.js

```diff
@@ -52,13 +52,13 @@ function toErrorSchema(errors) {
       }
       parent = parent[segment];
     }
-    if (Array.isArray(parent.__errors)) {
+    if (Array.isArray(parent.__errors) && message) {
```

Comment:
What is this code (changes made on lines 50-61) meant for?

Reply:
This counters the message-less paths inserted at L129-133 that are used to create empty `__errors` arrays at those nodes that don't have any errors when the flattened data is expanded to the hierarchical error schema object. Without this a couple tests failed:

https://github.com/mozilla-services/react-jsonschema-form/blob/aebfab9baa94d35176b7197ecf080e2fc1258dcf/test/validate_test.js#L488-L542

https://github.com/mozilla-services/react-jsonschema-form/blob/aebfab9baa94d35176b7197ecf080e2fc1258dcf/test/validate_test.js#L544-L575

It [seems to be intentional](https://github.com/mozilla-services/react-jsonschema-form/blob/aebfab9baa94d35176b7197ecf080e2fc1258dcf/src/validate.js#L93
) that `__errors` are inserted on every node regardless of errors.

It's been awhile so I'm not 100% sure how these transformations work anymore, I'll have a deeper look next week and document the data structures provided to and expected out of transformErrors and customValidate, including the custom error handling I'm implementing here.

Reply:
So you added some docs in this area recently, haven't had a chance to review yet myself but will try to next week